mllib/src/main/scala/org/apache/spark/ml/stat/FValueTest.scala

```diff
@@ -0,0 +1,111 @@
+   * @return DataFrame containing the test result for every feature against the label.
+   *         This DataFrame will contain a single Row with the following fields:
+   *          - `pValues: Vector`
+   *          - `degreesOfFreedom: Array[Int]`
+   *          - `fValues: Vector`
+   *         Each of these fields has one value per feature.
+   */
+  @Since("3.1.0")
+  def test(dataset: DataFrame, featuresCol: String, labelCol: String): DataFrame = {
+
+    val spark = dataset.sparkSession
+    import spark.implicits._
+
+    SchemaUtils.checkColumnType(dataset.schema, featuresCol, new VectorUDT)
+    SchemaUtils.checkNumericType(dataset.schema, labelCol)
+
+    val Row(xMeans: Vector, xStd: Vector, yMean: Double, yStd: Double, count: Long) = dataset
+      .select(Summarizer.metrics("mean", "std", "count").summary(col(featuresCol)).as("summary"),
+        avg(col(labelCol)).as("yMean"),
+        stddev(col(labelCol)).as("yStd"))
+      .select("summary.mean", "summary.std", "yMean", "yStd", "summary.count")
+      .first()
+
+    val labeledPointRdd = dataset.select(col("label").cast("double"), col("features"))
+      .as[(Double, Vector)].rdd
+
+    val numFeatures = xMeans.size
+    val numSamples = count
+    val degreeOfFreedom = numSamples.toInt - 2
+
+    // Use two pass equation Cov[X,Y] = E[(X - E[X]) * (Y - E[Y])] to compute covariance because
+    // one pass equation Cov[X,Y] = E[XY] - E[X]E[Y] is susceptible to catastrophic cancellation
+    // sumForCov = Sum(((Xi - Avg(X)) * ((Yi-Avg(Y)))
+    val sumForCov = labeledPointRdd.mapPartitions { iter =>
+      if (iter.hasNext) {
+        val array = Array.ofDim[Double](numFeatures)
+        while(iter.hasNext) {
+          val (label, features) = iter.next
+          val yDiff = label - yMean
+          if (yDiff != 0) {
+            features.iterator.zip(xMeans.iterator)
+              .foreach { case ((col, x), (_, xMean)) => array(col) += yDiff * (x - xMean) }
+          }
+        }
+        Iterator.single(array)
+      } else Iterator.empty
+    }.treeReduce { case (array1, array2) =>
+      var i = 0
+      while (i < numFeatures) {
+        array1(i) += array2(i)
+        i += 1
+      }
+      array1
+    }
+
+    val pValues = Array.ofDim[Double](numFeatures)
+    val degreesOfFreedoms = Array.ofDim[Int](numFeatures)
+    val fValues = Array.ofDim[Double](numFeatures)
+
+    val fd = new FDistribution(1, degreeOfFreedom)
+    for(i <- 0 until numFeatures) {
+      // Cov(X,Y) = Sum(((Xi - Avg(X)) * ((Yi-Avg(Y))) / (N-1)
+      val covariance = sumForCov (i) / (numSamples - 1)
+      val corr = covariance / (yStd * xStd(i))
+      val fValue = corr * corr / (1 - corr * corr) * degreeOfFreedom
+      pValues(i) = 1.0 - fd.cumulativeProbability(fValue)
+      degreesOfFreedoms(i) = degreeOfFreedom
+      fValues(i) = fValue
+    }
+
+    spark.createDataFrame(Seq((Vectors.dense(pValues), degreesOfFreedoms, Vectors.dense(fValues))))
```

Comment:
I think we need to add a private case class `FValueResult` like(`ChiSquareResult`&`KolmogorovSmirnovTestResult`), and use it to create final dataframe.